### PR TITLE
prl-tools: 12.2.1-41615 -> 17.1.4-51567

### LINKED ISSUE #31882: parallels-guest.nix is broken
## Issue description

Configuring NixOS to use Parallels Tools is impossible. 
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/parallels-guest.nix

### Steps to reproduce
1. Install NixOS as a Parallels guest VM on Mac
2. Add `hardware.parallels.enable = true;` to nix config and run `nixos-rebuild switch`
3. Expect the following error
```
***
Please, place Parallels Tools for Linux image into Nix store
using either
  nix-store --add-fixed sha256 prl-tools-lin.iso
or
  nix-prefetch-url file://path/to/prl-tools-lin.iso

***
```
4. Copy the `prl-tools-lin.iso` from the Mac /Applications folder into NixOS
5. As advised, run one of the above commands to add the .iso file
7. Re-run `nixos-rebuild switch` which (unexpectedly) repeats the same error in step 3

## Technical details

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.61, NixOS, 17.09.2075.ac35504065 (Hummingbird)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 1.11.15`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`
```

### LINKED ISSUE #79280: Cannot install hardware.parallels.package
**Describe the bug**
Take an error:

```
building Nix...
building the system configuration...
error: The option value `hardware.parallels.package' in `/nix/var/nix/profiles/per-user/root/channels/nixos/nixos/modules/virtualisation/parallels-guest.nix' is not of type `package'.
(use '--show-trace' to show detailed location information)
```

**To Reproduce**
1. Add to `configuration.nix`:
```
nixpkgs.config.allowUnfree = true;
hardware.parallels.enable = true;
```
2. `nixos-rebuild switch`

**Expected behavior**
Installed parallels tools

**Screenshots**

**Additional context**

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.
```
$ nix-shell -p nix-info --run "nix-info -m"
these paths will be fetched (41.05 MiB download, 174.14 MiB unpacked):
  /nix/store/0hr45a0pzlh51hhcgynmfjpzff9d3ddv-binutils-2.31.1
  /nix/store/24l5rfm2xhqkx96jvq0rg8pplb89b9hv-gcc-8.3.0
  /nix/store/28ad36blb1ar2lbccamxnwf9mwb3nn3h-expand-response-params
  /nix/store/2n8wj6qh69c5mkc87mhfa74qip5gb86p-binutils-wrapper-2.31.1
  /nix/store/8x3lspbpjlm5wpwg3zb2qiksrxblsnax-gcc-wrapper-8.3.0
  /nix/store/fldkzixl4ascrrg1cs3j229gn6b9xxiw-patchelf-0.9
  /nix/store/k7vxqz5y3h6f73i4nhy2dp539kxcfi6r-bash-interactive-4.4-p23-dev
  /nix/store/lrnpy7nv1rrigc4d62fiwr4wkryimzxr-linux-headers-4.19.16
  /nix/store/qwm593x9sxd48vz09s523qi4zywfhsks-glibc-2.27-dev
  /nix/store/v28wx4y2mq5y0i6r2n4yfw71lqvz7wxz-stdenv-linux
copying path '/nix/store/k7vxqz5y3h6f73i4nhy2dp539kxcfi6r-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/0hr45a0pzlh51hhcgynmfjpzff9d3ddv-binutils-2.31.1' from 'https://cache.nixos.org'...
copying path '/nix/store/28ad36blb1ar2lbccamxnwf9mwb3nn3h-expand-response-params' from 'https://cache.nixos.org'...
copying path '/nix/store/lrnpy7nv1rrigc4d62fiwr4wkryimzxr-linux-headers-4.19.16' from 'https://cache.nixos.org'...
copying path '/nix/store/fldkzixl4ascrrg1cs3j229gn6b9xxiw-patchelf-0.9' from 'https://cache.nixos.org'...
copying path '/nix/store/qwm593x9sxd48vz09s523qi4zywfhsks-glibc-2.27-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/2n8wj6qh69c5mkc87mhfa74qip5gb86p-binutils-wrapper-2.31.1' from 'https://cache.nixos.org'...
copying path '/nix/store/24l5rfm2xhqkx96jvq0rg8pplb89b9hv-gcc-8.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/8x3lspbpjlm5wpwg3zb2qiksrxblsnax-gcc-wrapper-8.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/v28wx4y2mq5y0i6r2n4yfw71lqvz7wxz-stdenv-linux' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.78, NixOS, 19.09.741.dbad7c7d59f (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```
Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### LINKED ISSUE #179117: Parallels guest doesn't work
### Describe the bug

`hardware.parallels.enable = true` doesn't enable parallels guest on kernel `5.15.47`.

### Steps To Reproduce
1. Add `hardware.parallels.enable = true`;
2. `nixos-rebuild switch`

### Expected behavior
Should work.

### Screenshots
```console
[catap@nixos:/etc/nixos]$ sudo nixos-rebuild switch
building Nix...
building the system configuration...
error: cannot coerce null to a string

       at /nix/var/nix/profiles/per-user/root/channels/nixos/nixos/modules/virtualisation/parallels-guest.nix:92:22:

           91|       serviceConfig = rec {
           92|         ExecStart = "${prl-tools}/sbin/prlfsmountd ${PIDFile}";
             |                      ^
           93|         ExecStartPre = "${pkgs.coreutils}/bin/mkdir -p /media";
(use '--show-trace' to show detailed location information)

[catap@nixos:/etc/nixos]$ uname -a
Linux nixos 5.15.47 #1-NixOS SMP Tue Jun 14 16:36:28 UTC 2022 x86_64 GNU/Linux

[catap@nixos:/etc/nixos]$ 
```

### Additional context

It requires kernel bigger than 4.15: https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/linux-kernels.nix#L427

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

### PR DESCRIPTION
**Pull request by catap**
###### Description of changes

Update prl-tools to the last version, tested that it works fine under Parallels 17.

It also:
 - fixes https://github.com/NixOS/nixpkgs/issues/179117,
 - fixes https://github.com/NixOS/nixpkgs/issues/31882,
 - and fixes https://github.com/NixOS/nixpkgs/issues/79280

This PR + https://github.com/NixOS/nixpkgs/pull/179487 and simple overlays:
```nix
    self: super: {
      systemd = super.systemd.overrideAttrs (old: {
        patches = old.patches ++ [
          (super.fetchpatch {
            url = "https://gist.githubusercontent.com/catap/be19980b57e01015ada609e448ca41b8/raw/ba7e61a07bddd59336a0519dfabed2551b5a0cbc/virt-fix-detection-of-Parallels-virtualization.patch";
            sha256 = "sha256-PyCUgioN3P/C+SzwYXLaIdsWuWyT1jJvYsTPFypASyw=";
          })
        ];
      });
    }
```

allows to install parallels tools on VM fully automatically :)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by wegank — APPROVED** [review]

File: pkgs/os-specific/linux/prl-tools/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/prl-tools/default.nix:122]:
```suggestion
    platforms = platforms.linux;
```

File: pkgs/os-specific/linux/prl-tools/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/prl-tools/default.nix:14]: Please use assert instead.

**Review by SuperSandro2000 — COMMENTED** [review]
Please squash the commits together for the package.

File: pkgs/os-specific/linux/prl-tools/default.nix
Comment (catap) [reply pkgs/os-specific/linux/prl-tools/default.nix:122]: it won't work without `with lib` I guess

Comment (catap) [reply pkgs/os-specific/linux/prl-tools/default.nix]: => removed all `lib.`

Comment (SuperSandro2000) [reply pkgs/os-specific/linux/prl-tools/default.nix]: we have with lib so this works

**Review by wegank — APPROVED** [review]

**Review by sandangel — APPROVED** [review]
thanks a lot for working on this

File: nixos/modules/virtualisation/parallels-guest.nix
Comment (flokli) [inline nixos/modules/virtualisation/parallels-guest.nix:7]: This is only used in one place. Please inline there.

File: nixos/modules/virtualisation/parallels-guest.nix
Comment (flokli) [inline nixos/modules/virtualisation/parallels-guest.nix:56]: please inline the aarch64 check here (and break this up in two lines)

File: nixos/modules/virtualisation/parallels-guest.nix
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/parallels-guest.nix:56]:
```suggestion
      ++ optional (pkgs.stdenv.hostPlatform.system == "aarch64-linux") "prl_notifier";
```

**Review by wegank — APPROVED** [review]

**Comment by catap** [comment]
@SuperSandro2000 may I ask you re-review one more time?

File: pkgs/os-specific/linux/prl-tools/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/prl-tools/default.nix:12]:
```suggestion
  version = "17.1.4-51567";
```

We can get prl_major like this:
```
nix-repl> lib.versions.major "17.1.4-51567"
"17"
```

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 may I ask you re-review one more time?

I don't really know anything about prl-tools so I am a bit hesitant with merging.

File: pkgs/os-specific/linux/prl-tools/default.nix
Comment (catap) [reply pkgs/os-specific/linux/prl-tools/default.nix]: @SuperSandro2000 thanks, fixed.

**Comment by catap** [comment]
@SuperSandro2000 it is working. I'm using it for a few weeks for now.

I assume that @wegank and @sandangel also had tested this code and enjoy it.

How can I help you to decide to merge it?

**Review by wegank — APPROVED** [review]

**Comment by negz** [comment]
FWIW I've been running this for a week per https://github.com/negz/nix/blob/be24cbb/hosts/mael/hardware-configuration.nix#L7.

The only issue I've noticed so far is that `prltimesync` isn't running - I notice if I start a Fedora VM that it's a child of `prltoolsd` along with a few other things. On my system `prltoolsd` has no children:

```
$ pstree 1460373
--= 1460373 root /nix/store/zbw072dym5mlnjg26zz1xk4cxsm6dr7g-prl-tools-17.1.4-51567/bin/prltoolsd -f

$ systemctl status prltoolsd.service
● prltoolsd.service - Parallels Tools Service
     Loaded: loaded (/etc/systemd/system/prltoolsd.service; enabled; vendor preset: enabled)
     Active: active (running) since Fri 2022-07-08 21:20:25 PDT; 1min 19s ago
   Main PID: 1460373 (prltoolsd)
         IP: 0B in, 0B out
         IO: 0B read, 4.0K written
      Tasks: 4 (limit: 19088)
     Memory: 392.0K
        CPU: 205ms
     CGroup: /system.slice/prltoolsd.service
             └─1460373 /nix/store/zbw072dym5mlnjg26zz1xk4cxsm6dr7g-prl-tools-17.1.4-51567/bin/prltoolsd -f

Jul 08 21:20:25 mael systemd[1]: Started Parallels Tools Service.
```

Also worth noting that I'm running without a graphical session so I'm not using/testing any wanted by `wantedBy = [ "graphical-session.target" ];`.

**Comment by catap** [comment]
@negz interesting, digging why.

**Comment by catap** [comment]
It has much more hardcoded paths. Will try to patch everything in a few days.

**Comment by catap** [comment]
@negz, @wegank can I ask you to check one more time?

Right now it keeps `prltimesync` running and also printers should work.

The only error / missed part that I found is: ` Error: prlusmd account manager error when find sessions type: can't open dir` but I have no idea how `prlusmd` should be used in terms of nix, do you?

**Comment by catap** [comment]
Just to confirm that copy-and-paste and changes screen size is working. Or at least at xorg + pantheon.

**Comment by negz** [comment]
@catap Thank you! I copied your changes over per https://github.com/negz/nix/commit/a0d9a599 and now I see `prltimesync` is a child of `prltoolsd`. FWIW I see `prlusmd` running without any errors, but I have no idea what it does.

```console
$ systemctl status prltoolsd.service
● prltoolsd.service - Parallels Tools Service
     Loaded: loaded (/etc/systemd/system/prltoolsd.service; enabled; vendor preset: enabled)
     Active: active (running) since Sat 2022-07-09 14:55:03 PDT; 1min 54s ago
   Main PID: 1725899 (.prltoolsd-wrap)
         IP: 700B in, 980B out
         IO: 2.4M read, 4.0K written
      Tasks: 9 (limit: 19088)
     Memory: 3.9M
        CPU: 420ms
     CGroup: /system.slice/prltoolsd.service
             ├─1725899 /nix/store/lv1qnh1fwd7py5x7qx2nhyawnadphfin-prl-tools-17.1.4-51567/bin/prltoolsd -f
             ├─1725915 ./prltimesync
             └─1725916 ./prlusmd
```

Also FYI I'm running this on an M1 Mac so I can confirm it's all working (at least the non-graphical stuff) on Linux aarch64:

```console
$ uname -a
Linux mael 5.18.7 #1-NixOS SMP Sat Jun 25 13:29:48 UTC 2022 aarch64 GNU/Linux
```

**Comment by sandangel** [comment]
> I don't really know anything about prl-tools so I am a bit hesitant with merging.

hi @SuperSandro2000 , I think it's impossible to know everything when maintaining this huge repository. What can we do to make you feel more confident before merging? What are the risks that you can think of if we merge this PR? Should we add some kind of testing?

**Comment by catap** [comment]
@sandangel to be honest I have no idea how to test this software outside of parallels environment :)

**Comment by SuperSandro2000** [comment]
> What are the risks that you can think of if we merge this PR?

The worst that can happen is that people write angry comments.

Testing for unfree packages is a bit tricky. Maybe we just merge this and go from there if something breaks.

**Comment by ylluminate** [comment]
Has anyone done any work with this on Parallels on Apple Silicon?

**Comment by sandangel** [comment]
I'm using this on M1 mac, work perfectly fine.

File: nixos/modules/virtualisation/parallels-guest.nix
Comment (wegank) [inline nixos/modules/virtualisation/parallels-guest.nix:90]: TODO: recheck

Comment (wegank) [inline nixos/modules/virtualisation/parallels-guest.nix:90]: TODO: recheck

**Comment by sandangel** [comment]
btw, is there a way to use the module and package from unstable, but still everything else under `release-22.05` ?

**Comment by catap** [comment]
> btw, is there a way to use the module and package from unstable, but still everything else under `release-22.05` ?

Nope, you can’t overlay a module.

**Comment by sandangel** [comment]
@SuperSandro2000 , hi, could you share the code? currently I already copied the module code into local folder. but I would like to remove that and somehow pull it from nix unstable branch.

**Comment by SuperSandro2000** [comment]
I think that should help you https://discourse.nixos.org/t/overriding-a-module/6464/3?u=sandro

**Comment by hgl** [comment]
@catap do you use Wayland? I tried this out for a Wayland setup using Parallels 17.1.4-51567 on an Intel Mac, encountered some issues. 

I tried two configurations:

### 1. Wayland with gdm as the display manager:

```nix
{ config, pkgs, ... }:

{
  imports = [ ./hardware-configuration.nix ];
  boot.loader.systemd-boot.enable = true;
  boot.loader.efi.canTouchEfiVariables = true;
  hardware.parallels.enable = true;
  networking.useDHCP = false;
  networking.interfaces.enp0s5.useDHCP = true;

  services.xserver = {
    enable = true;
    displayManager.gdm.enable = true;
    displayManager.autoLogin.enable = true;
    displayManager.autoLogin.user = "hgl";
  };
  programs.sway = {
    enable = true;
    extraSessionCommands = ''
      export WLR_NO_HARDWARE_CURSORS=1
    '';
  };

  users.mutableUsers = false;
  users.users.hgl = {
    isNormalUser = true;
    extraGroups = [ "wheel" ];
    hashedPassword = "<password>";
  };

  system.stateVersion = "22.11";
}
```
With this config, the mouse is very sluggish, trailing behind my real movement.

### 2. Wayland with greetd as the display manager:

```nix
{ config, pkgs, ... }:

{
  imports = [ ./hardware-configuration.nix ];
  boot.loader.systemd-boot.enable = true;
  boot.loader.efi.canTouchEfiVariables = true;
  hardware.parallels.enable = true;
  networking.useDHCP = false;
  networking.interfaces.enp0s5.useDHCP = true;

  services.greetd = {
    enable = true;
    settings = rec {
      initial_session = {
        command = "${pkgs.sway}/bin/sway";
        user = "hgl";
      };
      default_session = initial_session;
    };
  };
  programs.sway = {
    enable = true;
    extraSessionCommands = ''
      export WLR_NO_HARDWARE_CURSORS=1
    '';
  };

  users.mutableUsers = false;
  users.users.hgl = {
    isNormalUser = true;
    extraGroups = [ "wheel" ];
    hashedPassword = "<password>";
  };

  system.stateVersion = "22.11";
}
```
With this config, whenever my mouse moves into the VM window, the mouse disappears.

And in both cases, the clipboard is not shared between the host and guest.

Maybe I missed setting something?

**Comment by catap** [comment]
@hgl nope, I haven't use / try it :(

**Comment by hgl** [comment]
@sandangel Thanks. Switching to gnome did make all the issues go away.

**Pull request merged**
